File: channel-manager-java/src/main/java/com/channelmanager/java/exception/ApiException.java
package com.channelmanager.java.exception; // 예외 패키지

import org.springframework.http.HttpStatus; // HTTP 상태 코드 열거형

// API 예외의 기본 클래스
// 모든 비즈니스 예외는 이 클래스를 상속하여 HTTP 상태 코드를 함께 전달한다
// RuntimeException을 상속하므로 체크 예외가 아니다 (Reactive 체인에서 사용 편리)
// Kotlin에서는 open class로 선언하여 상속을 허용하지만,
// Java에서는 클래스가 기본적으로 상속 가능하므로 별도 키워드가 필요 없다
public class ApiException extends RuntimeException {

    private final HttpStatus status; // 이 예외에 대응하는 HTTP 상태 코드

    // 생성자: HTTP 상태 코드와 예외 메시지를 받는다
    // super(message)로 RuntimeException에 메시지를 전달한다
    public ApiException(HttpStatus status, String message) {
        super(message); // 부모 클래스에 메시지 전달
        this.status = status; // HTTP 상태 코드 저장
    }

    // HTTP 상태 코드 getter
    // Kotlin에서는 val status로 선언하면 getter가 자동 생성되지만,
    // Java에서는 명시적으로 getter를 작성해야 한다 (Lombok @Data를 쓰지 않는 이유: 상속 구조)
    public HttpStatus getStatus() {
        return status;
    }
}

File: channel-manager-kotlin/src/main/kotlin/com/channelmanager/kotlin/exception/ApiException.kt
package com.channelmanager.kotlin.exception // 예외 패키지

import org.springframework.http.HttpStatus // HTTP 상태 코드 열거형

// API 예외의 기본 클래스
// 모든 비즈니스 예외는 이 클래스를 상속하여 HTTP 상태 코드를 함께 전달한다
// RuntimeException을 상속하므로 체크 예외가 아니다 (Reactive 체인에서 사용 편리)
// sealed class가 아닌 open class를 사용하여 하위 클래스가 자유롭게 상속할 수 있도록 한다
open class ApiException(
    val status: HttpStatus,   // 이 예외에 대응하는 HTTP 상태 코드
    message: String           // 예외 메시지 (RuntimeException에 전달)
) : RuntimeException(message)

// 404 Not Found - 요청한 리소스를 찾을 수 없을 때 발생
// 예: 존재하지 않는 재고 ID로 조회, 존재하지 않는 객실 타입 ID
class NotFoundException(message: String) : ApiException(HttpStatus.NOT_FOUND, message)

// 400 Bad Request - 요청 데이터가 비즈니스 규칙에 맞지 않을 때 발생
// 예: 가용 수량이 음수, 종료일이 시작일보다 이전
class BadRequestException(message: String) : ApiException(HttpStatus.BAD_REQUEST, message)

// 409 Conflict - 중복 데이터 등 충돌이 발생했을 때 사용
// 예: 동일 (roomTypeId, stockDate) 조합의 재고가 이미 존재
class ConflictException(message: String) : ApiException(HttpStatus.CONFLICT, message)
